main/core/injector.ts

```diff
@@ -282,12 +344,199 @@ export class Injector implements IInjector {
      * Resets the injector back to its default state
      */
     public reset() {
-        this._isDestroyed = false;
         this.cache.clear();
         this.tokenCache.clear();
-        this.children = new Map<string, Injector>();
-        this.registrations.clear();
-        this._metrics.clear();
+        this.children.clear();
+        this._registrations.clear();
+        this.metrics.clear();
+    }
+
+    private throwRegistrationNotFound(constructorType: any, ancestry: any[]) {
+        throw new Error(
+            `Cannot construct Type '${constructorType.name}' with ancestry '${ancestry
+                .map(ancestor => ancestor.name)
+                .join(
+                    ' -> ',
+                )}' the type is either not decorated with @Injectable or injector.register was not called for the type and configuration has constructUndecoratedTypes set to false`,
+        );
+    }
+
+    private createInstance<T extends new (...args: any[]) => any>(
+        type: T,
+        updateCache: boolean = false,
+        options?: IConstructionOptions<T>,
+        ancestors: any[] = [],
+        injector: IInjector = globalReference[DI_ROOT_INJECTOR_KEY],
+    ): InstanceType<T> {
+        // Do our base checks to see if we are exceeding our depth limits
+        if (ancestors.length > injector.configuration.maxTreeDepth) {
+            throw new Error(
+                `Cannot construct Type '${(type as any).name}' with ancestry '${ancestors
+                    .map(ancestor => ancestor.name)
+                    .join(' -> ')}' as max tree depth has been reached`,
+            );
+        }
+
+        let instance: any;
+
+        const overrideParams = (options || {}).params || [];
+        const constructorParamTypes = getConstructorTypes(type);
+        // Note this call to construct param values introduces recursive behavior
+        const constructorParamValues = this.getConstructorParamValues<T>(
+            constructorParamTypes,
+            overrideParams,
+            injector,
+            ancestors,
+            type,
+        );
+
+        instance = new type(...constructorParamValues);
+
+        if (updateCache) {
+            injector.cache.update(type, instance);
+        }
+
+        return instance;
+    }
+
+    /**
+     * This method will resolve all the constructor args values taking into account the tokens
+     */
+    private getConstructorParamValues<T extends new (...args: any[]) => any>(
```

Comment:
Can't figure out why this is all new. We had this functionality already didn't we?

Reply:
It's just been refactored into a named method. The injector is self contained now meaning construction code is local to class in order to aid in passing, resolving parent content